File: frontend/src/api/events.ts
import api from './client'

export type EventType = 'running' | 'idle' | 'down' | 'maint'
export type MaintType = 'maint' | 'down' | 'idle'

export interface MachineEventOut {
  id: string
  machine_id: string
  event_type: EventType
  started_at: string
  ended_at: string | null
  quality_pct: number
  note: string | null
  created_at: string
  duration_min: number | null
  interventions: InterventionOut[]
}

export interface InterventionOut {
  id: string
  event_id: string
  machine_id: string
  cause: string | null
  action: string | null
  technician: string | null
  duration_min: number | null
  created_at: string
}

export interface PlannedMaintenanceOut {
  id: string
  machine_id: string
  planned_date: string
  duration_min: number
  maint_type: MaintType
  reason: string | null
  is_done: boolean
  created_at: string
}

export interface TimelineResponse {
  machine_id: string
  events: MachineEventOut[]
  planned: PlannedMaintenanceOut[]
}

export interface EventCreate {
  event_type: EventType
  started_at: string
  ended_at?: string
  quality_pct?: number
  note?: string
}

export interface PlannedCreate {
  planned_date: string
  duration_min: number
  maint_type?: MaintType
  reason?: string
}

export const eventsApi = {
  getTimeline: async (machineId: string, days = 14): Promise<TimelineResponse> => {
    const { data } = await api.get(`/machines/${machineId}/timeline?days=${days}`)
    return data
  },

  createEvent: async (machineId: string, payload: EventCreate): Promise<MachineEventOut> => {
    const { data } = await api.post(`/machines/${machineId}/events`, payload)
    return data
  },

  updateEvent: async (eventId: string, payload: Partial<EventCreate>): Promise<MachineEventOut> => {
    const { data } = await api.put(`/events/${eventId}`, payload)
    return data
  },

  deleteEvent: async (eventId: string): Promise<void> => {
    await api.delete(`/events/${eventId}`)
  },

  createPlanned: async (machineId: string, payload: PlannedCreate): Promise<PlannedMaintenanceOut> => {
    const { data } = await api.post(`/machines/${machineId}/planned-maintenance`, payload)
    return data
  },

  deletePlanned: async (pmId: string): Promise<void> => {
    await api.delete(`/planned-maintenance/${pmId}`)
  },
}

File: backend/app/schemas/events.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, field_validator

from app.models.events import EventType, MaintType


# ═══════════════════════════════════════════════════════
# MACHINE EVENTS
# ═══════════════════════════════════════════════════════

class MachineEventCreate(BaseModel):
    event_type: EventType
    started_at: datetime
    ended_at: Optional[datetime] = None
    quality_pct: int = 100
    note: Optional[str] = None

    @field_validator("quality_pct")
    @classmethod
    def clamp_quality(cls, v: int) -> int:
        return max(0, min(100, v))


class MachineEventUpdate(BaseModel):
    ended_at: Optional[datetime] = None
    quality_pct: Optional[int] = None
    note: Optional[str] = None


class InterventionOut(BaseModel):
    model_config = {"from_attributes": True}
    id: uuid.UUID
    event_id: uuid.UUID
    machine_id: uuid.UUID
    cause: Optional[str]
    action: Optional[str]
    technician: Optional[str]
    duration_min: Optional[int]
    created_at: datetime


class MachineEventOut(BaseModel):
    model_config = {"from_attributes": True}
    id: uuid.UUID
    machine_id: uuid.UUID
    event_type: EventType
    started_at: datetime
    ended_at: Optional[datetime]
    quality_pct: int
    note: Optional[str]
    created_at: datetime
    duration_min: Optional[int] = None
    interventions: list[InterventionOut] = []

    @classmethod
    def model_validate(cls, obj, **kwargs):
        instance = super().model_validate(obj, **kwargs)
        # Calculer duration_min depuis le modèle
        if hasattr(obj, 'duration_min'):
            instance.duration_min = obj.duration_min
        return instance


# ═══════════════════════════════════════════════════════
# INTERVENTIONS
# ═══════════════════════════════════════════════════════

class InterventionCreate(BaseModel):
    cause: Optional[str] = None
    action: Optional[str] = None
    technician: Optional[str] = None
    duration_min: Optional[int] = None


# ═══════════════════════════════════════════════════════
# PLANNED MAINTENANCE
# ═══════════════════════════════════════════════════════

class PlannedMaintenanceCreate(BaseModel):
    planned_date: datetime
    duration_min: int
    maint_type: MaintType = MaintType.maint
    reason: Optional[str] = None


class PlannedMaintenanceUpdate(BaseModel):
    planned_date: Optional[datetime] = None
    duration_min: Optional[int] = None
    maint_type: Optional[MaintType] = None
    reason: Optional[str] = None
    is_done: Optional[bool] = None


class PlannedMaintenanceOut(BaseModel):
    model_config = {"from_attributes": True}
    id: uuid.UUID
    machine_id: uuid.UUID
    planned_date: datetime
    duration_min: int
    maint_type: MaintType
    reason: Optional[str]
    is_done: bool
    created_at: datetime


# ═══════════════════════════════════════════════════════
# TIMELINE — réponse agrégée frontend
# ═══════════════════════════════════════════════════════

class TimelineResponse(BaseModel):
    machine_id: uuid.UUID
    events: list[MachineEventOut]
    planned: list[PlannedMaintenanceOut]
